src/utils/charts.js

```diff
@@ -0,0 +1,255 @@
+      actionsByTimeOfDay[NIGHT] += 1;
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(`actionHourOfDay ${actionHourOfDay} is undefined`);
+    }
+  });
+  return actionsByTimeOfDay;
+};
+
+// Takes object with {key: value} pairs of {timeOfDay: #-of-actions}
+// returns a date-sorted array in Recharts.js format
+export const formatActionsByTimeOfDay = (actionsByTimeOfDayObject) => {
+  const actionsByTimeOfDayArray = Object.entries(actionsByTimeOfDayObject);
+  return actionsByTimeOfDayArray.map(([timeOfDay, count]) => {
+    return {
+      timeOfDay,
+      count,
+    };
+  });
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {verb: %-of-actions}
+export const getActionsByVerb = (actions) => {
+  const totalActionsCount = actions.length;
+  const actionsByVerbCount = {};
+  actions.forEach((action) => {
+    if (!actionsByVerbCount[action.verb]) {
+      // if verb is still not in the actionsByVerb object, add it and initialize it to 1
+      actionsByVerbCount[action.verb] = 1;
+      // else increment existing count
+    } else {
+      actionsByVerbCount[action.verb] += 1;
+    }
+  });
+
+  // we want to return an object with key-value pairs of {verb: PERCENTAGE-of-total-actions}
+  // therefore need to convert count of actions to a percentage
+  // (1) convert actionsByVerbCount to an array of 2d arrays with [verb, count] via Object.entries
+  // (2) map onto [verb, count / totalActionsCount]
+  // (3) convert back into an object via Object.fromEntries
+  const actionsByVerbCountArray = Object.entries(actionsByVerbCount);
+  const actionsByVerbPercentageArray = actionsByVerbCountArray.map(
+    ([verb, count]) => [verb, count / totalActionsCount],
+  );
+  const actionsByVerbPercentage = Object.fromEntries(
+    actionsByVerbPercentageArray,
+  );
+
+  return actionsByVerbPercentage;
+};
+
+export const formatActionsByVerb = (actionsByVerbObject) => {
+  const actionsByVerbArray = Object.entries(actionsByVerbObject);
+
+  // capitalize verbs (entry[0][0]), convert 0.0x notation to x% and round to two decimal places (entry[0][1])
+  const formattedActionsByVerbArray = actionsByVerbArray
+    .map(([verb, ratio]) => [
+      _.capitalize(verb),
+      parseFloat((ratio * 100).toFixed(2)),
+    ])
+    .filter((entry) => entry[1] >= MIN_PERCENTAGE_TO_SHOW_VERB);
+
+  // add ['other', x%] to cover all verbs that are filtered out of the array
+  if (formattedActionsByVerbArray.length) {
+    formattedActionsByVerbArray.push([
+      OTHER_VERB,
+      // ensure that it is a number with two decimal places
+      parseFloat(
+        (
+          100 -
+          formattedActionsByVerbArray.reduce(
+            (acc, current) => acc + current[1],
+            0,
+          )
+        ).toFixed(2),
+      ),
+    ]);
+  }
+
+  // convert to recharts required format
+  return formattedActionsByVerbArray.map(([verb, percentage]) => {
```

Comment:
Just a remark as well: you can refactor it as `([verb, percentage]) => ({verb, percentage})`. The parenthesis allow to return object without errors.

Reply:
good call!